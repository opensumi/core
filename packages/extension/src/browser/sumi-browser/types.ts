import { IExtension } from '../..';
import { ProxyIdentifier } from '@opensumi/ide-connection/lib/common/rpcProtocol';
import { IDisposable, Uri } from '@opensumi/ide-core-common';
import { EditorComponentRenderMode } from '@opensumi/ide-editor/lib/browser';
import { Path } from '@opensumi/ide-core-common/lib/path';
import { ToolBarPosition } from '@opensumi/ide-toolbar/lib/browser';
import { ContextKeyExpr } from '@opensumi/monaco-editor-core/esm/vs/platform/contextkey/common/contextkey';

export interface ISumiBrowserContributions {
  [containerId: string]: {
    type?: 'replace' | 'add',
    view: IEditorViewContribution[] | ITabBarViewContribution[];
  } | undefined;
  left?: {
    type: 'replace' | 'add',
    view: ITabBarViewContribution[],
  };
  right?: {
    type: 'replace' | 'add',
    view: ITabBarViewContribution[],
  };
  bottom?: {
    type: 'replace' | 'add',
    view: ITabBarViewContribution[],
  };
  editor?: {
    view: IEditorViewContribution[];
  };
  editorSide?: {
    view: IEditorSideViewContribution[];
  };
  toolBar?: {
    position?: ToolBarPosition // @deprecated
    view: IToolBarViewContribution[];
  };
}

export interface IToolBarViewContribution {

  /**
   * id
   */
  id: string;
  /**
   * ToolBar 组件主体
   */
  component: React.FC;

  /**
   * 位置
   */
  position: ToolBarPosition;

  order?: number;

  weight?: number;

  description?: string;
}

export interface ITabBarViewContribution {

  /**
   * id
   */
  id: string;
  /**
   * Tabbar组件主体
   */
  component: React.FC;
  /**
   * 自定义titleComponent组件，仅支持插件注册container
   */
  titleComponent?: React.FC;

  /**
   * 内置icon名称
   */
  icon?: string;

  /**
   * 相对于插件路径的icon地址
   */
  iconPath?: string;

  /**
   * 用于激活的快捷键
   */
  keyBinding?: string;

  /**
   * 名称
   */
  title: string;

  /**
   * 自定义标题组件id
   */
  titleComponentId?: string;

  /**
   * 排序权重
   */
  priority?: number;

  /**
   * 禁止面板的resize功能
   */
  noResize?: boolean;

  /**
   * 是否全部展开
   */
  expanded?: boolean;
}

export interface IEditorViewContribution {

  /**
   * id
   */
  id: string;

  /**
   * 适配的scheme, 如果不填，默认为file协议
   */
  scheme?: string;

  /**
   * editor组件主体
   */
  component: React.FC;

  /**
   * 渲染方式
   */
  renderMode?: EditorComponentRenderMode;

  /**
   * 仅作用于file协议
   * 要处理的文件的后缀
   */
  fileExt?: string[];

  /**
   * 仅作用于file协议
   * 判断一个path是否要被处理
   * @deprecated
   */
  shouldPreview?: (path: Path) => boolean;

  /**
   * 判断一个uri是否要被处理(传入参数为vscode uri)
   * 如果不存在handles方法，则默认显示（file协议还要过shouldPreview和fileExt)
   */
  handles?: (uri: Uri) => boolean;

  /**
   * 如果这个资源有多个打开方式，这个会作为打开方式名称
   */
  title?: string;

  /**
   * 排序权重， 默认为10
   */
  priority?: number;

  /**
   * Tab名称，如果需要更复杂的名称Resolve，需要在 sumi node进程中注册ResourceProvider
   */
  tabTitle?: string;

   /**
   * 相对于插件路径的icon地址
   * 如果需要更复杂的图标Resolve，需要在 sumi node进程中注册ResourceProvider
   */
  tabIconPath?: string;
}

export interface IEditorSideViewContribution {
  /**
   * id
   */
  id: string;

  /**
  * editor组件主体
  */
  component: React.FC;

  /**
   * view 放置的位置，目前只支持 bottom
   */
  side: 'bottom';
  /**
   * When条件语句 https://code.visualstudio.com/api/references/when-clause-contexts#conditional-operators
   */
  when?: string | ContextKeyExpr;
}
export interface IRunTimeParams {
  getExtensionExtendService: (extension: IExtension, componentId: string) => {
    extendProtocol: {
      getProxy: (identifier: ProxyIdentifier<any>) => {
      node: any,
      worker: any,
      },
      set: <T>(identifier: ProxyIdentifier<T>, service: T) => void,
    },
    extendService: any,
  };
}

export abstract class AbstractSumiBrowserContributionRunner {

  constructor(protected extension: IExtension, protected contribution: ISumiBrowserContributions) {}

  abstract run(param: IRunTimeParams): IDisposable;

}
