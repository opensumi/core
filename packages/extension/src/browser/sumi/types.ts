import { IToolbarActionBtnStyle, IToolbarSelectStyle, IToolbarPopoverStyle } from '@opensumi/ide-core-browser';

export interface IToolbarActionBasicContribution {
  id: string;
  weight?: number;
  preferredPosition?: {
    location?: string,
    group?: string,
  };
  strictPosition?: {
    location: string,
    group: string,
  };
  description: string;
}

export interface IToolbarButtonContribution extends  IToolbarActionBasicContribution {
  type: 'button';
  command?: string;
  title: string;
  iconPath: string;
  iconMaskMode?: boolean;
  states?: {
    [key: string]: {
      title?: string,
      iconPath?: string,
      iconMaskMode?: boolean ;
    } & IToolbarActionBtnStyle,
  };
  defaultState?: string;
  // popover 元素的 component id
  popoverComponent?: string;

  popoverStyle?: IToolbarPopoverStyle;
}

export interface IToolbarSelectContribution<T = any> extends IToolbarActionBasicContribution {
  type: 'select';
  command?: string;
  options: {
    iconPath?: string,
    iconMaskMode?: boolean;
    label?: string,
    value: T
  }[];
  defaultValue: T;
  optionEqualityKey?: string;
  states?: {
    [key: string]: IToolbarSelectStyle,
  };
  defaultState?: string;
}
