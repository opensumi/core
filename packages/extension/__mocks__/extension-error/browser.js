(function webpackUniversalModuleDefinition(root, factory) {
	if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory(require("React"));
	else if(typeof define === 'function' && define.amd)
		define(["React"], factory);
	else if(typeof exports === 'object')
		exports["extend-browser-aligenie-plugin"] = factory(require("React"));
	else
		root["extend-browser-aligenie-plugin"] = factory(root["React"]);
})({}, function(__WEBPACK_EXTERNAL_MODULE__0__) {
return /******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, { enumerable: true, get: getter });
/******/ 		}
/******/ 	};
/******/
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = function(exports) {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/
/******/ 	// create a fake namespace object
/******/ 	// mode & 1: value is a module id, require it
/******/ 	// mode & 2: merge all properties of value into the ns
/******/ 	// mode & 4: return value when already ns object
/******/ 	// mode & 8|1: behave like require
/******/ 	__webpack_require__.t = function(value, mode) {
/******/ 		if(mode & 1) value = __webpack_require__(value);
/******/ 		if(mode & 8) return value;
/******/ 		if((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;
/******/ 		var ns = Object.create(null);
/******/ 		__webpack_require__.r(ns);
/******/ 		Object.defineProperty(ns, 'default', { enumerable: true, value: value });
/******/ 		if(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));
/******/ 		return ns;
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";
/******/
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = 1);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ (function(module, exports) {

module.exports = __WEBPACK_EXTERNAL_MODULE__0__;

/***/ }),
/* 1 */
/***/ (function(module, exports, __webpack_require__) {

module.exports = __webpack_require__(2);


/***/ }),
/* 2 */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
// ESM COMPAT FLAG
__webpack_require__.r(__webpack_exports__);

// EXTERNAL MODULE: external "React"
var external_React_ = __webpack_require__(0);

// CONCATENATED MODULE: ./src/extend/browser/mockplugin.tsx

const PluginsPanel = () => {
    function handleClick() {
        //
    }
    return external_React_["createElement"]("div", { onClick: handleClick }, "mock panel");
};

const PluginsEditor = () => {
  return external_React_["createElement"]("div", {} , "mock editor");
}

const PluginsToolBar = () => {
  return external_React_["createElement"]("div", {} , "mock editor");
}
// CONCATENATED MODULE: ./src/extend/browser/index.ts

/* harmony default export */ var browser = __webpack_exports__["default"] = ({
    left: {
        type: "add",
        component: [
            {
                id: "PluginsPanel",
                iconPath: './icons/entry.svg',
                priority: 10,
                panel: PluginsPanel
            }
        ]
    },
    editor: {
      type: "add",
      component: [
          {
              id: "PluginsEditor",
              panel: PluginsEditor
          }
      ]
  },
  toolBar: {
    type: "add",
    component: [
        {
            id: "PluginsToolBar",
            panel: PluginsToolBar
        }
    ]
},
});


/***/ })
/******/ ]);
});
