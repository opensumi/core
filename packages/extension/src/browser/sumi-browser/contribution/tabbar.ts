import { IRunTimeParams, AbstractSumiBrowserContributionRunner, ITabBarViewContribution } from '../types';
import { IDisposable, Disposable } from '@opensumi/ide-core-common';
import { Injectable, Autowired } from '@opensumi/di';
import { IMainLayoutService } from '@opensumi/ide-main-layout';
import { getIcon } from '@opensumi/ide-core-browser';
import { IIconService } from '@opensumi/ide-theme';

const SUPPORT_LOCATION = ['left', 'right', 'bottom', 'editor', 'toolBar'];

@Injectable({ multiple: true })
export class TabbarBrowserContributionRunner extends AbstractSumiBrowserContributionRunner {
  @Autowired(IMainLayoutService)
  layoutService: IMainLayoutService;

  @Autowired(IIconService)
  iconService: IIconService;

  run(params: IRunTimeParams): IDisposable {
    const disposer = new Disposable();
    if (this.contribution.left) {
      this.contribution.left.view.forEach((view) => {
        disposer.addDispose(this.registerTabBar(view, params, this.contribution.left?.type, 'left'));
      });
    }

    if (this.contribution.right) {
      this.contribution.right.view.forEach((view) => {
        disposer.addDispose(this.registerTabBar(view, params, this.contribution.right?.type, 'right'));
      });
    }

    if (this.contribution.bottom) {
      this.contribution.bottom.view.forEach((view) => {
        disposer.addDispose(this.registerTabBar(view, params, this.contribution.bottom?.type, 'bottom'));
      });
    }

    Object.keys(this.contribution).forEach((location) => {
      if (SUPPORT_LOCATION.indexOf(location) === -1) {
        (this.contribution[location]?.view as ITabBarViewContribution[]).forEach((view) => {
          disposer.addDispose(this.registerTabBar(view, params, 'replace'));
        });
      }
    });

    return disposer;
  }

  registerTabBar(
    view: ITabBarViewContribution,
    runtimeParams: IRunTimeParams,
    kind: 'add' | 'replace' = 'add',
    position?: 'left' | 'right' | 'bottom',
  ): IDisposable {
    const { extendProtocol, extendService } = runtimeParams.getExtensionExtendService(this.extension, view.id);
    const containerId = `${this.extension.id}:${view.id}`;
    const initialProps = {
      kaitianExtendService: extendService,
      kaitianExtendSet: extendProtocol,
    };
    this.layoutService.viewReady.promise.then(() => {
      if (kind === 'add') {
        const { component, titleComponent } = view;
        this.layoutService.collectTabbarComponent(
          [
            {
              id: containerId,
            },
          ],
          {
            ...view,
            iconClass: view.icon
              ? getIcon(view.icon)
              : view.iconPath
              ? this.iconService.fromIcon(this.extension.path, view.iconPath)
              : '',
            containerId,
            component,
            titleComponent,
            titleProps: initialProps,
            initialProps,
            fromExtension: true,
          },
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          position!,
        );
      } else {
        this.layoutService.replaceViewComponent(view, initialProps);
        view.titleComponent &&
          this.layoutService.getTabbarHandler(containerId)?.setTitleComponent(view.titleComponent, initialProps);
      }
    });

    return {
      dispose: () => {
        const componentHandler = this.layoutService.getTabbarHandler(containerId);
        if (componentHandler) {
          componentHandler.dispose();
        }
      },
    };
  }
}
