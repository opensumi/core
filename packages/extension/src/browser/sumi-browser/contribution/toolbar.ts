import { IRunTimeParams, AbstractSumiBrowserContributionRunner } from '../types';
import { IDisposable, Disposable, ILogger } from '@opensumi/ide-core-common';
import { Injectable, Autowired, INJECTOR_TOKEN, Injector } from '@opensumi/di';
import { IMainLayoutService } from '@opensumi/ide-main-layout';
import { IIconService } from '@opensumi/ide-theme';
import { IToolBarViewService, ToolBarPosition } from '@opensumi/ide-toolbar/lib/browser';

@Injectable({ multiple: true })
export class ToolBarBrowserContributionRunner extends AbstractSumiBrowserContributionRunner {
  @Autowired(IMainLayoutService)
  layoutService: IMainLayoutService;

  @Autowired(IToolBarViewService)
  toolBarViewService: IToolBarViewService;

  @Autowired(IIconService)
  iconService: IIconService;

  @Autowired(ILogger)
  logger: ILogger;

  @Autowired(INJECTOR_TOKEN)
  injector: Injector;

  run(param: IRunTimeParams): IDisposable {
    const disposer = new Disposable();
    if (!this.injector.creatorMap.has(IToolBarViewService)) {
      this.logger.warn('没有找到 toolbarViewService');
      return disposer;
    }

    if (this.contribution.toolBar) {
      this.contribution.toolBar.view.forEach((view) => {
        const { extendProtocol, extendService } = param.getExtensionExtendService(this.extension, view.id);
        const disposable = this.toolBarViewService.registerToolBarElement({
          type: 'component',
          component: view.component as React.ComponentType,
          position: view.position || this.contribution.toolBar?.position || ToolBarPosition.LEFT,
          initialProps: {
            kaitianExtendService: extendService,
            kaitianExtendSet: extendProtocol,
          },
          description: view.description,
          order: view.order,
          weight: view.weight,
        });
        if (disposable) {
          disposer.addDispose(disposable);
        }
      });
    }

    return disposer;
  }
}
