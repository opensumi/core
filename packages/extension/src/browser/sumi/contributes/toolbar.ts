import { Injectable, Autowired } from '@opensumi/di';
import { IJSONSchema, IToolbarRegistry } from '@opensumi/ide-core-browser';
import { LifeCyclePhase } from '@opensumi/ide-core-browser/lib/bootstrap/lifecycle.service';
import { toolbar } from '@opensumi/ide-core-browser/lib/extensions/schema/toolbar';

import { VSCodeContributePoint, Contributes, LifeCycle } from '../../../common';
import { AbstractExtInstanceManagementService } from '../../types';
import { KaitianExtensionToolbarService } from '../main.thread.toolbar';
import { IToolbarButtonContribution, IToolbarSelectContribution, IToolbarActionBasicContribution } from '../types';

export interface KtToolbarSchema {
  actions?: Array<IToolbarButtonContribution | IToolbarSelectContribution>;
  groups?: Array<{
    id: string;
    preferredLocation?: string;
    weight?: number;
  }>;
}

@Injectable()
@Contributes('toolbar')
@LifeCycle(LifeCyclePhase.Starting)
export class ToolbarContributionPoint extends VSCodeContributePoint<KtToolbarSchema> {
  @Autowired()
  private readonly kaitianExtToolbarService: KaitianExtensionToolbarService;

  @Autowired(IToolbarRegistry)
  private readonly toolbarRegistry: IToolbarRegistry;

  @Autowired(AbstractExtInstanceManagementService)
  protected readonly extensionManageService: AbstractExtInstanceManagementService;

  static schema: IJSONSchema = toolbar.schema;

  private toLocalized<T extends IToolbarActionBasicContribution>(action: T, props: string[], extensionId: string): T {
    return props.reduce((pre, cur) => {
      if (pre[cur]) {
        pre[cur] = this.getLocalizeFromNlsJSON(pre[cur], extensionId);
      }
      return pre;
    }, action);
  }

  contribute() {
    for (const contrib of this.contributesMap) {
      const { extensionId, contributes } = contrib;
      const extension = this.extensionManageService.getExtensionInstanceByExtId(extensionId);
      if (!extension) {
        continue;
      }

      if (contributes.groups) {
        for (const group of contributes.groups) {
          this.addDispose(
            this.toolbarRegistry.registerToolbarActionGroup({
              id: group.id,
              preferredLocation: group.preferredLocation,
              weight: group.weight,
            }),
          );
        }
      }
      if (contributes.actions) {
        for (const toolbarAction of contributes.actions) {
          if (toolbarAction.type === 'button') {
            this.addDispose(
              this.kaitianExtToolbarService.registerToolbarButton(
                extensionId,
                extension.path,
                this.toLocalized(toolbarAction, ['title'], extensionId),
              ),
            );
          } else if (toolbarAction.type === 'select') {
            this.addDispose(
              this.kaitianExtToolbarService.registerToolbarSelect(
                extensionId,
                extension.path,
                this.toLocalized(toolbarAction, ['description'], extensionId),
              ),
            );
          }
        }
      }
    }
  }
}
