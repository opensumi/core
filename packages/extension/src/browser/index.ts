import { Injectable, Provider } from '@opensumi/di';
import { BrowserModule } from '@opensumi/ide-core-browser';
import { DebugSessionContributionRegistry } from '@opensumi/ide-debug/lib/browser/debug-session-contribution';
import { IDebugServer } from '@opensumi/ide-debug/lib/common';
import { FileSearchServicePath } from '@opensumi/ide-file-search/lib/common';

import {
  AbstractExtensionManagementService,
  ExtensionHostProfilerServicePath,
  ExtensionNodeServiceServerPath,
  ExtensionService,
  IExtCommandManagement,
  IRequireInterceptorService,
  RequireInterceptorContribution,
  RequireInterceptorService,
} from '../common';
import {
  AbstractNodeExtProcessService,
  AbstractViewExtProcessService,
  AbstractWorkerExtProcessService,
} from '../common/extension.service';
import {
  IMainThreadExtenderService,
  MainThreadExtenderContribution,
  MainThreadExtenderService,
} from '../common/main.thread.extender';

import { ActivationEventServiceImpl } from './activation.service';
import { ExtCommandManagementImpl as ExtCommandManagementImpl } from './extension-command-management';
import { ExtInstanceManagementService } from './extension-instance-management';
import { ExtensionManagementService } from './extension-management.service';
import { NodeExtProcessService } from './extension-node.service';
import { ViewExtProcessService } from './extension-view.service';
import { WorkerExtProcessService } from './extension-worker.service';
import { ExtensionClientAppContribution, ExtensionCommandContribution } from './extension.contribution';
import { ExtensionServiceImpl } from './extension.service';
import { BrowserRequireInterceptorContribution } from './require-interceptor.contribution';
import { SumiContributionsService, SumiContributionsServiceToken } from './sumi/contributes';
import { AbstractExtInstanceManagementService, IActivationEventService } from './types';
import { ExtensionDebugService, ExtensionDebugSessionContributionRegistry } from './vscode/api/debug';
import { VSCodeContributesService, VSCodeContributesServiceToken } from './vscode/contributes';

@Injectable()
export class ExtensionModule extends BrowserModule {
  contributionProvider = [RequireInterceptorContribution, MainThreadExtenderContribution];
  providers: Provider[] = [
    {
      token: ExtensionService,
      useClass: ExtensionServiceImpl,
    },
    {
      token: IDebugServer,
      useClass: ExtensionDebugService,
      override: true,
    },
    {
      token: DebugSessionContributionRegistry,
      useClass: ExtensionDebugSessionContributionRegistry,
      override: true,
    },
    {
      token: IActivationEventService,
      useClass: ActivationEventServiceImpl,
    },
    {
      token: IExtCommandManagement,
      useClass: ExtCommandManagementImpl,
    },
    {
      token: AbstractExtInstanceManagementService,
      useClass: ExtInstanceManagementService,
    },
    {
      token: AbstractExtensionManagementService,
      useClass: ExtensionManagementService,
    },
    {
      token: AbstractNodeExtProcessService,
      useClass: NodeExtProcessService,
    },
    {
      token: AbstractWorkerExtProcessService,
      useClass: WorkerExtProcessService,
    },
    {
      token: AbstractViewExtProcessService,
      useClass: ViewExtProcessService,
    },
    {
      token: IRequireInterceptorService,
      useClass: RequireInterceptorService,
    },
    {
      token: VSCodeContributesServiceToken,
      useClass: VSCodeContributesService,
    },
    {
      token: SumiContributionsServiceToken,
      useClass: SumiContributionsService,
    },
    {
      token: IMainThreadExtenderService,
      useClass: MainThreadExtenderService,
    },
    ExtensionCommandContribution,
    ExtensionClientAppContribution,
    BrowserRequireInterceptorContribution,
  ];

  backServices = [
    {
      servicePath: ExtensionNodeServiceServerPath,
      clientToken: ExtensionService,
    },
    {
      servicePath: FileSearchServicePath,
    },
    {
      servicePath: ExtensionHostProfilerServicePath,
    },
  ];
}
