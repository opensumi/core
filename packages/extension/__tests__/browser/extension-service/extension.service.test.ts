/* eslint-disable @typescript-eslint/ban-ts-comment */
import ReactDom from 'react-dom';
import {
  ExtensionService,
  IExtCommandManagement,
  AbstractExtensionManagementService,
  IRequireInterceptorService,
} from '../../../src/common';
import { AbstractExtInstanceManagementService } from '../../../src/browser/types';
import { MockInjector } from '../../../../../tools/dev-tool/src/mock-injector';
import {
  CommandRegistryImpl,
  CommandRegistry,
  IPreferenceSettingsService,
  PreferenceScope,
  KeybindingRegistryImpl,
  KeybindingRegistry,
} from '@opensumi/ide-core-browser';
import { IThemeService, getColorRegistry } from '@opensumi/ide-theme/lib/common';
import { IMenuRegistry, MenuRegistryImpl, IMenuItem } from '@opensumi/ide-core-browser/src/menu/next';
import { IMainLayoutService } from '@opensumi/ide-main-layout';
import { LayoutService } from '@opensumi/ide-main-layout/lib/browser/layout.service';
import { PreferenceSettingsService } from '@opensumi/ide-preferences/lib/browser/preference-settings.service';
import { WorkbenchThemeService } from '@opensumi/ide-theme/lib/browser/workbench.theme.service';
import { IToolbarRegistry } from '@opensumi/ide-core-browser/lib/toolbar';
import { NextToolbarRegistryImpl } from '@opensumi/ide-core-browser/src/toolbar/toolbar.registry';
import { IActivationEventService, ExtensionBeforeActivateEvent } from '@opensumi/ide-extension/lib/browser/types';
import { TabbarService } from '@opensumi/ide-main-layout/lib/browser/tabbar/tabbar.service';
import { MOCK_EXTENSIONS, setupExtensionServiceInjector } from './extension-service-mock-helper';

describe('Extension service', () => {
  let extensionService: ExtensionService;
  let extCommandManagement: IExtCommandManagement;
  let extInstanceManagementService: AbstractExtInstanceManagementService;
  let extensionManagementService: AbstractExtensionManagementService;
  let injector: MockInjector;

  beforeAll(() => {
    injector = setupExtensionServiceInjector();
    injector.get(IMainLayoutService).viewReady.resolve();
    extensionService = injector.get(ExtensionService);
    extCommandManagement = injector.get(IExtCommandManagement);
    extInstanceManagementService = injector.get(AbstractExtInstanceManagementService);
    extensionManagementService = injector.get(AbstractExtensionManagementService);
  });

  describe('activate', () => {
    it('should activate extension service.', async (done) => {
      await extensionService.activate();
      done();
    });

    it('emit event before activate', async (done) => {
      // @ts-ignore
      extensionService.eventBus.on(ExtensionBeforeActivateEvent, () => {
        done();
      });

      // @ts-ignore
      await extensionService.doActivate();
    });

    it('emit onStartupFinished activationEvent after activate', async (done) => {
      const activationEventService = injector.get<IActivationEventService>(IActivationEventService);
      activationEventService.onEvent('onStartupFinished', () => {
        done();
      });
      // @ts-ignore
      await extensionService.doActivate();
    });
  });

  describe('get extension', () => {
    it.skip('should return all mock extensions', async () => {
      const exts = await extInstanceManagementService.getExtensionInstances();
      expect(exts).toEqual(MOCK_EXTENSIONS);
    });

    it('should return all mock extensions JSON', async () => {
      const jsons = await extensionManagementService.getAllExtensionJson();
      expect(jsons).toEqual(MOCK_EXTENSIONS.map((e) => e.toJSON()));
    });

    it('should return specified extension props', async () => {
      const extensionMetadata = await extensionManagementService.getExtensionProps(MOCK_EXTENSIONS[0].path, {
        readme: './README.md',
      });
      expect(extensionMetadata?.extraMetadata).toEqual({ readme: './README.md' });
    });

    it('should return extension by extensionId', async () => {
      const extension = extensionManagementService.getExtensionByExtId('test.sumi-extension');
      expect(extension?.extensionId).toBe(MOCK_EXTENSIONS[0].extensionId);
    });
  });

  describe('extension status sync', () => {
    it('should return false when extension is not running', async () => {
      const extension = await extensionManagementService.getExtensionByPath(MOCK_EXTENSIONS[0].path);
      expect(extension?.activated).toBe(false);
    });
  });

  describe('activate extension', () => {
    it('should activate mock browser extension without ext process', async (done) => {
      await extensionService.activeExtension(MOCK_EXTENSIONS[0]);
      const layoutService: IMainLayoutService = injector.get(IMainLayoutService);
      const tabbarService: TabbarService = layoutService.getTabbarService('left');
      const containerInfo = tabbarService.getContainer('test.sumi-extension:Leftview');
      expect(containerInfo?.options?.titleComponent).toBeDefined();
      expect(containerInfo?.options?.titleProps).toBeDefined();
      done();
      // setTimeout(() => {
      // }, 1000);
    });

    it('extension should not repeated activation', async () => {
      const extInstances = await extInstanceManagementService.getExtensionInstances();
      expect(extInstances).toHaveLength(1);
      await extensionManagementService.postChangedExtension(false, MOCK_EXTENSIONS[0].realPath);
      const postExtInstances = await extInstanceManagementService.getExtensionInstances();
      expect(postExtInstances).toHaveLength(1);
    });
  });

  describe('extension contributes', () => {
    it('should register toolbar actions via new toolbar action contribution point', () => {
      const toolbarRegistry: IToolbarRegistry = injector.get(IToolbarRegistry);
      (toolbarRegistry as NextToolbarRegistryImpl).init();
      const groups = toolbarRegistry.getActionGroups('default');
      expect(groups?.length).toBe(1);
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      expect(toolbarRegistry.getToolbarActions({ location: 'default', group: groups![0].id })?.actions?.length).toBe(1);
    });

    it('should register shadow command via command contribution point', () => {
      const commandRegistry: CommandRegistryImpl = injector.get(CommandRegistry);
      expect(commandRegistry.getCommand('HelloKaitian')).toBeDefined();
    });

    it('should register menus in editor/title and editor/context position', (done) => {
      const newMenuRegistry: MenuRegistryImpl = injector.get(IMenuRegistry);
      const contextMenu = newMenuRegistry.getMenuItems('editor/context');
      expect(contextMenu.length).toBe(1);
      expect((contextMenu[0] as IMenuItem).command).toBe('HelloKaitian');
      const actionMenu = newMenuRegistry.getMenuItems('editor/title');
      expect(actionMenu.length).toBe(1);
      expect(actionMenu.findIndex((item) => (item as IMenuItem).command === 'HelloKaitian')).toBeGreaterThan(-1);
      done();
    });

    it('should register viewContainer in activityBar', (done) => {
      const layoutService: LayoutService = injector.get(IMainLayoutService);
      const handler = layoutService.getTabbarHandler('package-explorer');
      expect(handler).toBeDefined();
      const holdHandler = layoutService.getTabbarHandler('hold-container');
      expect(holdHandler).toBeUndefined();
      done();
    });

    it('should register extension configuration', (done) => {
      const preferenceSettingsService: PreferenceSettingsService = injector.get(IPreferenceSettingsService);
      const preferences = preferenceSettingsService.getSections('extension', PreferenceScope.Default);
      expect(preferences.length).toBe(1);
      expect(preferences[0].title).toBe('Mock Extension Config');
      done();
    });

    it('should register browserView', (done) => {
      const layoutService: LayoutService = injector.get(IMainLayoutService);
      const tabbar = layoutService.getTabbarHandler('test.sumi-extension:KaitianViewContribute');
      expect(tabbar).toBeDefined();
      expect(tabbar?.containerId).toBe('test.sumi-extension:KaitianViewContribute');
      done();
    });

    it('should register browserView', (done) => {
      const layoutService: IMainLayoutService = injector.get(IMainLayoutService);
      const tabbar = layoutService.getTabbarHandler('test.sumi-extension:KaitianViewContribute');
      expect(tabbar).toBeDefined();
      done();
    });

    it('should register keybinding for HelloKaitian command', (done) => {
      const keyBinding: KeybindingRegistryImpl = injector.get(KeybindingRegistry);
      const commandKeyBindings = keyBinding.getKeybindingsForCommand('HelloKaitian');
      expect(commandKeyBindings.length).toBe(1);
      expect(typeof commandKeyBindings[0].keybinding).toBe('string');
      done();
    });

    it('should register mock color', async (done) => {
      const themeService: WorkbenchThemeService = injector.get(IThemeService);
      const colorRegister = getColorRegistry();
      const theme = await themeService.getCurrentTheme();
      const color = colorRegister.resolveDefaultColor('mock.superstatus.error', theme);
      expect(color).toBeDefined();
      expect(color?.toString()).toBe('#ff004f');
      done();
    });
  });

  describe('extension host commands', () => {
    it("should define a command in 'node' host.", async (done) => {
      const commandId = 'mock_command';
      const disposable = extCommandManagement.registerExtensionCommandEnv(commandId, 'node');
      const env = extCommandManagement.getExtensionCommandEnv(commandId);
      expect(env).toBe('node');
      disposable.dispose();
      expect(extCommandManagement.getExtensionCommandEnv(commandId)).toBe(undefined);
      done();
    });
  });

  describe('load browser require interceptor contribution', () => {
    it('should get ReactDOM interceptor', async () => {
      // @ts-ignore
      await extensionService.doActivate();
      const requireInterceptorService: IRequireInterceptorService = injector.get(IRequireInterceptorService);
      const interceptor = requireInterceptorService.getRequireInterceptor('ReactDOM');
      const result = interceptor?.load({});
      expect(result).toBe(ReactDom);
    });
  });
});
