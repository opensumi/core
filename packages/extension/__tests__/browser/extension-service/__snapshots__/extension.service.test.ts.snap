// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extension service load browser require interceptor contribution should get ReactDOM interceptor 1`] = `
{
  "__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED": {
    "Events": [
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
    ],
    "usingClientEntryPoint": false,
  },
  "createPortal": [Function],
  "createRoot": [Function],
  "findDOMNode": [Function],
  "flushSync": [Function],
  "hydrate": [Function],
  "hydrateRoot": [Function],
  "render": [Function],
  "unmountComponentAtNode": [Function],
  "unstable_batchedUpdates": [Function],
  "unstable_renderSubtreeIntoContainer": [Function],
  "version": "18.2.0",
}
`;
